### PR TITLE
[export] Serialize symbolic values

### PR DESCRIPTION
**Pull request by angelayi**
* Modified the SymInt schema to also store the hint of the SymInt if it is represented as a symbol so that when we reconstruct the SymInt, the hint will also exist on the node.
* GraphModuleDeserializer.deserialize now also optionally map of symbol names to range.

ReplaceSymSizeOpPass should not be needed after https://github.com/pytorch/pytorch/pull/103107 lands

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103273

### CODE REVIEW

File: torch/_export/serde/serialize.py
Comment (avikchaudhuri) [inline torch/_export/serde/serialize.py:630]: how did this pass lint, lol...don't we insist on trailing commas?

**Review by avikchaudhuri — CHANGES_REQUESTED** [review]

**Comment by avikchaudhuri** [comment]
Requesting changes mainly because I'm unsure of the motivation behind adding extra params to graph module deserializer.

File: torch/_export/serde/serialize.py
Comment (angelayi) [reply torch/_export/serde/serialize.py:630]: I don't think the OSS linter insists on having trailing commas

File: test/export/test_serialize.py
Comment (angelayi) [reply test/export/test_serialize.py:190]: If they are tensors then I have to use `torch.allclose` to check equality.

File: torch/_export/serde/serialize.py
Comment (angelayi) [reply torch/_export/serde/serialize.py:737]: The `deserialize_range_constraints` function is called before deserializing the graph module, and I had it populate the shape_env with the symbols that appeared in those constraints first. 

This is under the assumption that most symbols that appear in the graph will also appear in the constraints (I want to say all symbols? but I'm not 100% sure on that although I don't have a counterexample). So I can first create all of the symbols in the shape_env, and then pass the `shape_env` and `symbol_name_to_symbol` list to the `GraphModuleDeserializer`. 

We need to populate `shape_env` with the ranges before tracing because if there's some data-dependent operation going on, like using a symint as a shape, deserializing the tensor metadata through the `torch.empty_strided` call will check on the ranges of the symint.

The `symbol_name_to_symbol` is also handy because when we're deserializing symint expressions, the `sympy.sympify` function uses it to map the same expressions to the same symbol (ex. if we have `symbol_name_to_symbol = {"s0", s0}`, whenever we see `"s0"` this will point to the same `s0`, and if we see something like `"7*s0"` this will also point to the same symbol `s0`.

File: test/export/test_serialize.py
Comment (avikchaudhuri) [reply test/export/test_serialize.py:190]: Yeah I got that, just commenting on the double `isinstance` check.

File: torch/_export/serde/serialize.py
Comment (avikchaudhuri) [reply torch/_export/serde/serialize.py:737]: Ah, populated by side effect: missed that.

I think `constraints` only contain input dimension symbols, right? Whereas `shape_env` can contain data-dependent ones, I suppose.

I suppose `shape_env` better contain all symbols appearing in the graph. These should be a superset of things in `constraints`, I think. And `symbol_name_to_symbol` should have all symbols in the graph as well...?

Is there a way to assert on this?

Also, I still don't fully understand whether you actually expect external users to call graph module deserializer directly. Why not make it private, remove the optional / default part, etc.?

**Review by avikchaudhuri — APPROVED** [review]

File: torch/_export/serde/schema.py
Comment (zhxchen17) [inline torch/_export/serde/schema.py:75]: Why we don't go with the side table as discussed before? Are you going to change this in a later PR?

File: torch/_export/serde/serialize.py
Comment (zhxchen17) [inline torch/_export/serde/serialize.py:691]: I guess you are handling unbacked symint in a later PR?

**Review by zhxchen17 — COMMENTED** [review]
some initial comments

File: torch/_export/serde/serialize.py
Comment (zhxchen17) [inline torch/_export/serde/serialize.py:322]:
```suggestion
def _int_to_sympy_int(val) -> sympy.Expr:
```

**Review by zhxchen17 — COMMENTED** [review]
I think we're good to go after 1 more iteration.

File: torch/_export/serde/serialize.py
Comment (angelayi) [reply torch/_export/serde/serialize.py:709]: I can't because the constrain API doesn't take in a symbol as input 😅 but it uses a symbol to set the range...it's easier if I just set it directly...

File: torch/fx/experimental/symbolic_shapes.py
Comment (zhxchen17) [inline torch/fx/experimental/symbolic_shapes.py:302]: nit, we already have "experimental" in the module path, no need for underscore here.
```suggestion
def constrain_symbol_range(shape_env, s: sympy.Symbol, min: int, max: int):
```

**Review by zhxchen17 — APPROVED** [review]
let's go

File: torch/fx/experimental/symbolic_shapes.py
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:302]: Eh, if we know this is going to be private, why not keep the underscore? One day symbolic shapes will no longer be experimental, and we don't have to do additional cleanup then...

Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:302]: Anyway, don't have to block on this comment, idc so much... :)

Comment (zhxchen17) [reply torch/fx/experimental/symbolic_shapes.py:302]: I guess most of the functions in this file is meant to be private... yet they don't got underscore, so my point is mostly about consistency... but yeah don't feel blocked on this.

**Pull request closed without merging**
